### ISSUE TITLE
auth policy with custom header checks break (at least) push tokens

### ISSUE THREAD

**Issue by jmozd** [privacyidea/privacyidea#2514]
### Top-level intent
After creating policies that use custom HTTP headers, for me, authentication via push tokens started to fail. I chased this down to reported exceptions within policy handling:

a policy ("policyX") is defined, relying on a custom header (active condition comparing a header against a given value). Sample purpose of policyX is to inititate a push token request for a specific PAM scenario

the client sends the "/validate/check" POST request including the header, traverses the policy chain and policyX works as expected

i.e. the push token reply is coming in ("/ttype/push" GET request) and traverses the policy chain, but does not have the custom header

when checking "policyX" against that "/ttype/push" request, the missing HTTP header is reported as an error, an exception is thrown and further processing seems to stop

This PR changes the behavior of _policy_matches_request_header_condition() to only log a warning and to return as if the comparison on the header failed.

The comparator parameter of _policy_matches_request_header_condition() was not (yet) included in the decision. Returning "true" for "inverse matches" (!in, !contains,...) would possibly create a different semantic, but of course this can be subject of discussion.

### Steps to reproduce

1. create authentication policy using custom header, asking for push token
2. issue authentication request

### Expected outcome

The push token should pick up the request and once confirmed, the authentication should happen.

### Actual outcome

The push token never asks the user for confirmation, even refreshing the push token device does not show the request.
The authentication request times out.
What did actually happen?

### Configuration

* **privacyIDEA version: 3.4.1 (it'd be cool if this was part of the "system documentation")
* **Installation method: pip install privacyidea
* **Python version: Python 3.6.10 (it'd be cool if this was part of the "system documentation")
* **Operating system: OpenSUSE Leap 15.2
* **Webserver: apache2-2.4.43-lp152.1.1.x86_64
* **Token database: MariaDB database back-end

### Log file

When the httpd server receives the request "GET /ttype/push HTTP/1.1"  778 "-" "Dart/2.10 (dart:io)", PI's log file shows:

[140479983216384][ERROR][privacyidea.lib.policy:788] Policy 'policyX' has conditions on headers 'application', but http header is not available. This should not happen.

and it seems the policy processing stops here.

I tracked this error message down and demoted it to a "warning", returning as if the header actually exists, but does not match. Once I put that change into place, processing completed and worked as expected.

Created PR https://github.com/privacyidea/privacyidea/pull/2509 with the according change.

**Comment by cornelinux** [comment]
It was a design decision to fail a request, if it does not contain information about the required condition.
The argument was, that if we do not have the information, we can not say if the policy should apply or not.

The correct way,then would be to fix the request.

**Comment by jmozd** [comment]
The practical impact of that design decision can be pretty severe in more complex environments. Would you consider adding a comperator "existsAndEqual" a suitable solution to the situation? I might be able to provide an according PR.

**Comment by cornelinux** [comment]
You are totally right about the severity. But the problem arises with other extended conditions. Because you could change the outcome by simply removing some comparing values and this would result in unexpected behaviour.
And as even bigger problem this might result in a behaviour you would not be aware of, that this is a problem. The admin would *think* he configured everything nicely but he gets "wrong" policies applied without ever notificing it.

(We had lengthy internal discussions about this roughly a year ago, when we added this).
So we decided to go the hard, but clear way. Since the admin would be award of failing requests and thus take action to solve the logical issues.

If you would add "existsAndEqual" you would have to add such for all comporators.
We need to think of this if we think of a generic approach:

Take the "userinfo". A policy should apply if the attribute "mobile" of the user starts with "+123". 
So what should happen, if there is no attribute mobile? We can not decide. It gets mind boggling when you reverse the condition:
The policy should apply, if the attribute "mobile" of the user does not start with "+123".

If "mobile" does not exist, how should privacyIDEA be certain what to do.
So we recommend: rather take care that all your users have an attribute "mobile" - even if it is empty.

If you would go for a generic checkbox with the extended attributes - should it be

* [ ] apply this policy if the key does not exist
* [ ] do not apply this policy if the key does not exist
* [ ] treat a non existing key as empty string

Which would be very different outcomes and often not the outcome you would probably want.

**Comment by cornelinux** [comment]
Why can't you add the corresponding header in the request?

**Comment by jmozd** [comment]
> Why can't you add the corresponding header in the request?

Because it's the PI Android app sending the request :)

**Comment by jmozd** [comment]
Let me give you a description of what I'm trying to achieve - maybe I took a wrong turn in the early design stage and there's a better way to handle this.

I was trying to create a setup where I secure SSH and sudo on Linux hosts, both services are using PAM and I have created an according working version of the Python authenticator to handle push tokens for auth and session stage.

- if a user is authenticating via SSH private key token and the target host is internal, accept the access
- if a user is authenticating via SSH private key token and the target host is external, require MFA (currently via push token)
- if a user is authenticating via SSH and password, require MFA (currently via push token)
- if "sudo" is called, require (push) token confirmation from the user (configurable whether initiating or target user of sudo), but apply sudo rulesets first.

For this, I had created policies:

- authentication: ssh token login (no further filter rules during test)
- authentication: push token authentication
- authentication: push token authentication - but only if custom header "application" is set to "pam/sudo"

The users in question have (some of) their tokens (types ssh and push) assigned to my test host, for application "ssh".

The PAM setup of the test host is:

- ssh: "auth" using PI via Python script, setting "application" to "pam/ssh"
- sudo: "session" using PI via python script, setting "application" to "pam/sudo"

Without my PR, incoming "/ttype/push" GET requests from the PI Android app (after the authentication call by the PAM session authhandler) will fail the checks. Setting the client IP address range is easy for the "internal" hosts because of the limited corporate IP address range used for internal clients, but external hosts would need to be listed explicitly - all, for any rule checking that header. Which appears to be an administrative nightmare.

**Comment by cornelinux** [comment]
Ah, I read "dart". You are using the beta app, right?

**Comment by jmozd** [comment]
> Ah, I read "dart". You are using the beta app, right?

correct  - v3.0.7

**Comment by cornelinux** [comment]
I think the quickest approach would be to add the header "application" in the privacyIDEA Authenticator App.
This should send

    application: privavcyidea-authenticator

I will open an issue over there.

**Comment by jelinaunger** [comment]
Done in #4346
